plugins/Delay/DelayEffect.cpp

```diff
@@ -82,25 +82,38 @@ bool DelayEffect::processAudioBuffer( sampleFrame* buf, const fpp_t frames )
 		return( false );
 	}
 	double outSum = 0.0;
+	const float sr = Engine::mixer()->processingSampleRate();
 	const float d = dryLevel();
 	const float w = wetLevel();
-	const float length = m_delayControls.m_delayTimeModel.value() * Engine::mixer()->processingSampleRate();
-	const float amplitude = m_delayControls.m_lfoAmountModel.value() * Engine::mixer()->processingSampleRate();
-	m_lfo->setFrequency( 1.0 / m_delayControls.m_lfoTimeModel.value() );
-	m_delay->setFeedback( m_delayControls.m_feedbackModel.value() );
+	float length = m_delayControls.m_delayTimeModel.value();
+	float amplitude = m_delayControls.m_lfoAmountModel.value() * sr;
+	float lfoTime = 1.0 / m_delayControls.m_lfoTimeModel.value();
+	float feedback =  m_delayControls.m_feedbackModel.value();
 	sample_t dryS[2];
 	float lPeak = 0.0;
 	float rPeak = 0.0;
+	ValueBuffer *lengthBuffer = m_delayControls.m_delayTimeModel.valueBuffer();
+	ValueBuffer *feedbackBuffer = m_delayControls.m_feedbackModel.valueBuffer();
+	ValueBuffer *lfoTimeBuffer = m_delayControls.m_lfoTimeModel.valueBuffer();
+	ValueBuffer *lfoAmountBuffer = m_delayControls.m_lfoAmountModel.valueBuffer();
 	if( m_delayControls.m_outGainModel.isValueChanged() )
 	{
 		m_outGain = dbvToAmp( m_delayControls.m_outGainModel.value() );
 	}
 	for( fpp_t f = 0; f < frames; ++f )
 	{
-		m_currentLength = linearInterpolate( length, m_currentLength, 0.9999 );
 		dryS[0] = buf[f][0];
 		dryS[1] = buf[f][1];
-		m_delay->setLength( ( float )m_currentLength + ( amplitude * ( float )m_lfo->tick() ) );
+		length = lengthBuffer ? lengthBuffer->values()[ f ] : length;
```

Comment:
You could optimize this a bit. In amplifier, I used separate for loops for cases with sample-exact data and without... however, since you have 4 s.ex.controls here, that'd create 16 separate for loops (for each combination of s.ex/regular) which might be overkill and a bit messy, readability-wise.

So here's an idea. How about using a pointer for the value, so that before the loop, you can set it to point at either the first index of the buffer, or at a single variable if no buffer is available. Then, set increment variables for each model - 1 for s.ex, 0 for normal, and use them to increment the pointers in the loop. So in s.ex. case the pointer gets incremented accross the buffer, in regular case it gets incremented with 0 and stays on the variable.

For example, for the length variable:
```
float length = m_delayControls.m_delayTimeModel.value();
ValueBuffer *lengthBuffer = m_delayControls.m_delayTimeModel.valueBuffer();
int lengthInc = lengthBuffer ? 1 : 0;
float *lengthPtr = lengthBuffer ? &( lengthBuffer->values()[ 0 ] ) : &length;

for( fpp_t f = 0; f < frames; ++f )
{
// here just use *lengthPtr instead of length
// and at the end of the loop:
lengthPtr += lengthInc;
}
```
Then just apply the same for the other controls.